Firebase Integration

saeedbashir opened the pull request:
### Description

[MA-2980](https://openedx.atlassian.net/browse/MA-2980)
[MA-2977](https://openedx.atlassian.net/browse/MA-2977)

Add a description of your changes here.

### Notes
I have followed firebase [wiki](https://openedx.atlassian.net/wiki/display/MA/Firebase+Eventing+Strategy) for formatting keys and parameters. I will add 'GoogleService-Info' to git ignore before merging.
### How to test this PR

### Reviewers
If you've been tagged for review, please "Start a review" with the Github GUI. 
- Code review: @BenjiLee , @danialzahid94

BenjiLee commented:
Ugh, so many files changed :<

BenjiLee commented on Source/GoogleService-Info.plist:
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Add this file to .gitignore

BenjiLee replied:
add this file to .gitignore and delete this file

saeedbashir replied:
Sure.

BenjiLee commented:
@saeedbashir Could you add us to your firebase console?

BenjiLee commented:
Done with pass. Just the one gitignore addition and removal of google-services file.

danialzahid94 commented on Podfile:
```diff
@@ -18,6 +18,7 @@ pod 'FBSDKCoreKit', '~> 4.6'
 pod 'FBSDKLoginKit', '~> 4.6'
 pod 'Parse', '~> 1.7'
 pod 'Smartling.i18n', '~> 1.0'
+pod 'Firebase/Core'
```
Should we rather specify a version for this pod? That way, we would only explicitly be upgrading to a newer version, and won't be running into any unexpected changes before that.

saeedbashir commented:
@BenjiLee I have added you and Scott. There are some test logs, ignore them.

BenjiLee commented:
@saeedbashir I haven't received an invite email, could you invite again?

saeedbashir commented:
@BenjiLee Firebase wasn't sent email on adding to project. Could you please login and check?

BenjiLee commented:
Whats the app name?

saeedbashir commented:
edx-mobile-iOS

BenjiLee commented:
@saeedbashir I see it!

saeedbashir commented:
@danialzahid94 @BenjiLee Ready for review.

danialzahid94 commented on Source/FirebaseAnalyticsTracker.swift:
```diff
@@ -0,0 +1,155 @@
+        }
+        
+        // Firebase only supports 36 characters for parameter value
+        if formattedValue.characters.count > 36 {
```
Let's maybe keep a constant for the max count. So that it's a single change whenever we have to update this.

BenjiLee replied:
Google is well aware of their small character limit so it's possible they might increase this in the future.

danialzahid94 commented on Source/FirebaseAnalyticsTracker.swift:
```diff
@@ -0,0 +1,155 @@
+            parts[course] = components[1]
+            parts[run] = components[2]
+        }
+        else if components.count <= 1 {
```
Can we move this check above the main condition. So that the insertion code is there just once.

danialzahid94 commented on Source/FirebaseAnalyticsTracker.swift:
```diff
@@ -0,0 +1,155 @@
+        return string
+    }
+    
+    private func formateParamValue(value: String)-> String {
```
Typo: It would be `formatParamValue`

saeedbashir commented on Source/FirebaseAnalyticsTracker.swift:
```diff
@@ -0,0 +1,155 @@
+        return value
+    }
+    
+    private func canAddParameter(key: String) -> Bool {
```
Code is more readable by moving checks to a proper function name.

saeedbashir commented:
@danialzahid94  Ready for another pass.

BenjiLee commented on Podfile.lock:
```diff
@@ -1,60 +1,78 @@
-  - Parse (1.12.0):
-    - Bolts/Tasks (~> 1.5)
+  - OHHTTPStubs/OHPathHelpers (4.8.0)
+  - Parse (1.14.2):
```
We should probably remove this pod. Not sure if there any dependencies on this, but there shouldn't be any need to update this.

BenjiLee replied:
> Are you talking about Parse?

Yep. We don't use Parse Notifications AFAIK since they announce their were ending their service a year ago.

saeedbashir replied:
Currently our code base is using Parse so we can't remove it whiten the scope of this PR.

BenjiLee replied:
> Currently our code base is using Parse so we can't remove it whiten the scope of this PR.

Agreed, but there should be no need to update this pod.

saeedbashir replied:
Pods were being updated because of major pod version.

BenjiLee commented on Source/SegmentAnalyticsTracker.swift:
```diff
@@ -59,6 +60,11 @@ class SegmentAnalyticsTracker : NSObject, OEXAnalyticsTracker {
         
         SEGAnalytics.sharedAnalytics().track(event.displayName, properties: info)
+        
+        // WARNING: add GoogleService-Info file to git ignore and remove debug argument
```
Could you clarify this warning? Is it a warning to the developers? GoogleServices-Info should already be in the .gitignore file. How does the debug argument affect this?

saeedbashir replied:
This is for my self. File is not added to git ignore because if you guys want to run the code so you can. I will add it to git ignore before merging.

BenjiLee replied:
Got it. Wasn't sure if WARNING was another form of TODO, FIXME, or ERROR

BenjiLee commented:
Done with pass

saeedbashir commented on Source/FirebaseAnalyticsTracker.swift:
```diff
@@ -0,0 +1,150 @@
+        }
+        
+        if components.count == componentsCount {
+            parts[org] = components[0]
```
I have made changes for this but didn't found it on wiki.

saeedbashir commented:
@BenjiLee  Ready for another pass.

BenjiLee commented on Source/FirebaseAnalyticsTracker.swift:
```diff
@@ -123,7 +125,8 @@ class FirebaseAnalyticsTracker: NSObject {
         
         if components.count == componentsCount {
-            parts[org] = components[0]
+            // droping prefix
```
`droping` -> `dropping`

BenjiLee commented on Podfile:
```diff
@@ -18,6 +18,7 @@ pod 'FBSDKCoreKit', '~> 4.6'
 pod 'FBSDKLoginKit', '~> 4.6'
 pod 'Parse', '~> 1.7'
 pod 'Smartling.i18n', '~> 1.0'
+pod 'Firebase/Core', '= 3.9.0'
```
So... apparently they just raised the character limit in version 3.10. Let's update this to the latest.

BenjiLee replied:
We will need to revisit our strategy.

danialzahid94 commented on Source/FirebaseAnalyticsTracker.swift:
```diff
@@ -0,0 +1,153 @@
+            if value.isKindOfClass(NSDictionary) {
+                formatParamatersForFirebase(value as! [String: NSObject], formattedParams: &formattedParams)
+            }
+            else if canAddParameter(key) {
```
Even if we keep this as a separate method, I think we can use it's negation, and then add a `continue`. This will allow us to have one less conditional nesting.

saeedbashir commented:
@BenjiLee I have updated firebase version to 3.11.0 from 3.9.0. As now we have 100 characters for parameter so I have updated code accordingly. Now I'm not removing 'edx.bi.' from BI events and settings from 'settings.video.fullscreen'. Let me know if you want me to remove code of course_id, block_id and module_id parsing as well.

saeedbashir commented:
@BenjiLee @danialzahid94 Ready for another pass.

BenjiLee commented:
Squash and merge.

BenjiLee commented:
@saeedbashir I made a note on the wiki about block_id. https://openedx.atlassian.net/wiki/pages/viewpage.action?pageId=150241307

BenjiLee commented:
\o/
